src/Disks/IDiskTransaction.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <boost/noncopyable.hpp>
+#include <Disks/IDisk.h>
+
+namespace DB
+{
+
+struct RemoveRequest
+{
+    std::string path;
+    bool if_exists = false;
+
+    explicit RemoveRequest(std::string path_, bool if_exists_ = false)
+        : path(std::move(path_)), if_exists(std::move(if_exists_))
+    {
+    }
+};
+
+using RemoveBatchRequest = std::vector<RemoveRequest>;
+
+/// Simple interface batch execution of write disk operations.
+/// Method are almost equal to disk methods.
+struct IDiskTransaction : private boost::noncopyable
+{
+public:
+    /// Tries to commit all accumulated operations simultaneously.
+    /// If something fails rollback and throw exception.
+    virtual void commit() = 0;
+
+    virtual ~IDiskTransaction() = default;
+
+    /// Create directory.
+    virtual void createDirectory(const std::string & path) = 0;
+
+    /// Create directory and all parent directories if necessary.
+    virtual void createDirectories(const std::string & path) = 0;
+
+    /// Remove all files from the directory. Directories are not removed.
+    virtual void clearDirectory(const std::string & path) = 0;
+
+    /// Move directory from `from_path` to `to_path`.
+    virtual void moveDirectory(const std::string & from_path, const std::string & to_path) = 0;
+
+    virtual void moveFile(const String & from_path, const String & to_path) = 0;
+
+    virtual void createFile(const String & path) = 0;
+
+    /// Move the file from `from_path` to `to_path`.
+    /// If a file with `to_path` path already exists, it will be replaced.
+    virtual void replaceFile(const std::string & from_path, const std::string & to_path) = 0;
+
+    /// Copy file `from_file_path` to `to_file_path` located at `to_disk`.
+    virtual void copyFile(const std::string & from_file_path, const std::string & to_file_path) = 0;
```

Comment:
Yes, it's intended. We need to copy something to another disk only in case of moves between disks. Also, another disk can possibly work with another metadata storage, so it'll be impossible to achieve any transactional atomicity here.